database/interface.py

```diff
@@ -146,6 +146,27 @@ def delete_harvest_source(self, source_id):
         source = self.db.get(HarvestSource, source_id)
         if source is None:
             return "Harvest source not found"
+
+        # Delete packages in the CKAN associated with the same harvest source.
+        records = (
+            self.db.query(HarvestRecord)
+            .filter_by(harvest_source_id=source_id)
+            .all()
+        )
+        ckan_ids = [record.ckan_id for record in records]
+        if ckan_ids:
+            ckan = RemoteCKAN(
+                os.getenv("CKAN_API_URL"), 
+                apikey=os.getenv("CKAN_API_TOKEN")
+            )
+            for pkg_id in ckan_ids:
+                try:
+                    ckan.action.dataset_purge(id=pkg_id)
```

Comment:
i believe dataset_purge returns None when executed but can we increment a counter to indicate how many were purged without exception? ideally that number would be the same as the number of records deleted from the harvest_table after cascade. resource: https://stackoverflow.com/questions/8954116/sqlalchemy-get-rows-affected-by-bulk-delete

Reply:
> ideally that number would be the same as the number of records deleted from the harvest_table after cascade

Is there a way to extract that record count from a foreign key cascade delete as opposed to an explicit delete as mentioned in the SO article above?

Reply:
using [triggers](https://stackoverflow.com/questions/10124077/how-can-i-get-id-of-cascade-deleted-or-marked-for-delete-items-in-sqlalchemy)? i feel like triggers would have a noticeable impact on db performance so i'm immediately hesitant

Reply:
I think we should write up a ticket to encompass this. It's only something we're going to be doing very rarely so db performance will be less of an issue, but we would want to ensure that the operation completes successfully.

Definitely a candidate for ongoing load testing. Does deleting 20k records cripple the admin app, and if so, for how long? Would it make more sense to run a special task in the runner?

Some more on triggers:
- https://docs.sqlalchemy.org/en/14/core/ddl.html#controlling-ddl-sequences
- https://www.postgresql.org/docs/current/sql-createtrigger.html

Reply:
added check for unsuccessfully dataset_purge. Also confirm no records remains after delete harvest_source.